### DISCUSSION TITLE
I can't download UDEMY videos.

### DISCUSSION THREAD

**Discussion by Nereplameri** [aclap-dev/video-downloadhelper#1306] [Help and Q&A]
### Your question

No matter what course I try, no matter what quality, I can no longer download videos from UDEMY.
I checked other discussions on the forum. Some time has passed. Has it been fixed? Or is it just me?

![Ekran Görüntüsü - 2024-09-27 16-16-36](https://github.com/user-attachments/assets/5183dd44-91e9-426f-9c3d-6ec8d15dccce)

### Addon & CoApp version and details

version: 9.1.0.44
target: mozilla
channel: stable
lang: tr
coapp: "checking"
license: {"unneeded":true}
platform: x86-64 linux
UA: Mozilla/5.0 (X11; Linux x86_64; rv:130.0) Gecko/20100101 Firefox/130.0
{
    "used_history_button": true,
    "last_download_directory": {
        "__serializer_tag": "option_some",
        "__serializer_value": {
            "__serializer_tag": "primitive",
            "__serializer_value": "/home/alperen/dwhelper"
        }
    },
    "never_show_no_incognito_msg_again": true,
    "view_options": {
        "all_tabs": false,
        "low_quality": false,
        "sort_by_status": true,
        "sort_reverse": false,
        "show_button_clean": true,
        "show_button_clean_all": false,
        "show_button_convert_local": false,
        "hide_downloaded": false
    },
    "open_count_store": 3,
    "successfull_dl": 51
}

**Comment by GrampaWildWilly** [comment]
I do not work for VDH.  I'm just another user.

![#01](https://github.com/user-attachments/assets/31619b3f-07c8-4184-b15e-e273edccef8d)

This has been discussed so many times on here.  You should have done a search of this forum before you posted.  The content you are trying to download is protected by DRM.  That's what I'm showing you on your own image.  VDH has never, does not now, & will never properly download content protected by DRM.  DRM is expressly intended to prevent downloading of copyrighted material.  Breaking this protection would be illegal.  When you visit pages on that site, look at what Firefox is telling you.  If it shows the DRM icon I pointed out in your image, forget about downloading that content.  It is my understanding that some content on Udemy (& Hotmart & other education web sites) is protected by DRM & some is not.  It appears to be at the whim of the author.  What you are reporting is correct behavior of VDH, not a bug.

**Reply by devinschumacher** [reply]
then why does their website say they support downloading from Udemy? https://www.downloadhelper.net/asite/udemy

**Comment by sortedcord** [comment]
How come its protected by DRM yet it can be recorded by a screen capture tool like OBS? I thought that it was not possible to record DRM protected streams with obs... Not to mention the fact that you can play it in picture-in-picture mode which is also not possible for drm streams? This might as well come from my lack of knowledge and understanding of how drm works..

**Reply by Mattmjs** [reply]
Supposedly OBS could record DRM video but then they changed it so that even recording doesn't work. Maybe you have an older
version of the software. The idea of DRM is to allow the video to play online even in PiP mode but prevent anyone from
keeping a copy of it locally on your computer. The video player decrypts the stream so that the viewer can obviously watch it.
But if you attempt to download it, it's still encrypted therefore you get a result like this. Totally unusable file.

Recording works because it's like using a cassette tape to record a song directly from the radio station. It's captures the video as an
unencrypted stream like on the webpage. I've read somewhere online that hardware capture cards can also record video.

![drm video](https://github.com/user-attachments/assets/14544b3a-90ab-4ebe-adc8-620fdab32f61)

**Reply by GrampaWildWilly** [reply]
My understanding is that it's more about the hardware than it is about the software (OBS).  As I understand it, DRM decoding is implemented in a combination of circuitry scattered across the CPU chip, the motherboard chipset, & the video adapter chipset.  But if you have an old enough system, OBS still works to record even DRM protected content because the old hardware didn't know about DRM.  I could record DRM content on my old Windows 7 system, not because it was Windows 7 & not because it was older releases of OBS, but because it was an old computer, about 9 years old when I turned it off for the last time about 1 year ago.  That's when I put my fancy shmancy new computer into use with Windows 11 & suddenly I couldn't screen record anything with DRM.  My new hardware conforms to the new DRM protocols & in box standard configuration, I can't record content protected by DRM.  Actually, the new DRM protocols aren't particularly new.  They're just newer than the antique hardware I used to have.  I call it progress in reverse.

But I recently decided I wanted to see whether I could hook my set-top cable TV box directly to my computer in order to record TV shows that are available as online streams but protected by DRM.  Some web searches turned up a YouTube video by a guy named Wheezy.  That should be enough of a clue for you to find his stuff.  He's actually a gamer & his videos are all about game consoles (Xbox, Wii, Playstation, etc.) & capture devices.  His instructions talk about stripping off HDCP so you can record game play on your computer.  HDCP is kind of an equivalent thing to DRM for HDTV signals.  I read between the lines of what Wheezy explained & did something a bit simpler to connect my STB to my computer.  It worked as I had hoped.

But then I engaged in the always dangerous pastime known as thinking.  I rearranged things so my STB was no longer connected to my computer.  But I am now able to record online content, livestreams & not live content, protected by DRM.  This was not my original target functionality.  It was a surprise when I discovered I could do this.  I'm sure you can believe I'm quite happy about it.

You need at a minimum 2 pieces of hardware that you probably don't have.  I bought mine on Amazon.  The first is an HDMI splitter.  I got it for about $50.  This is the device that strips off HDCP.  I got the one Wheezy recommended.  The second bit of hardware is a video capture device.  I got it for about $18.  Wheezy, being the total gamer geek, has an expensive internal capture device that installs in an expansion slot on the motherboard.  (Plus a fancy switch, something I don't need.)  I got a cheap little thing.  It's more like a dongle.  It takes HDMI in on one end & puts its output to a USB connector on the other end.  It has some limitations on highest resolution & best audio but it's good enough for what I want to do.  If you want to capture resolutions of 4K & higher, if you want to capture surround audio, you'll probably need one of the more expensive items.  I believe there is an external dongle-type device now that does the higher end stuff.  I suppose you can get whatever suits your needs.  You might also need some HDMI interconnect cables.  That depends on what you happen to have lying around.

My configuration is like this.

![#01](https://github.com/user-attachments/assets/450294c9-ef0e-4993-afc3-69c4ccc71e47)

I'm not much of a draftsman so that drawing is perhaps a bit clumsy.  The NVidia adapter is inside my system case, with its output ports on the back of the system case.  The USB port is also on the back of the system case.  Everything else is outside the system case.

When you plug in the capture device, Windows 11 recognizes it as a capture device.  You don't need to install any drivers.  But you do need to supply some kind of software to read it.  That software is OBS.

That's all I'm going to tell you.  The rest is for you to do what I did: experiment.

I'm not sure I totally understand why this works.  I think the reasoning goes like this.  Any content protected by DRM has been decoded by the time it reaches the output ports of the NVidia adapter.  So the signal going to the HDMI splitter has no DRM protection on it.  Routing one of the outputs of the splitter back into the computer via the capture device allows decoded audio-video signal to arrive at the capture device.  At that point, OBS doesn't need to do anything special to record it.  I do not believe the splitter is stripping off DRM encoding.  It strips off HDCP encoding, but that is not present in the situation I'm describing here.  DRM encoding is present.  That is not HDCP encoding.  They're different things.  But the DRM protection is removed inside the system so that what escapes from the back of the NVidia adapter is no longer encoded.  This is how I see things.  I could be wrong.  I'm not an expert in these things.  I'm just fitting an explanation to what I've observed.

I have 2 monitors.  Any content protected by DRM I send to my second monitor, which is my TV.  I then tell OBS to record what's on my TV.  If you have only 1 monitor, I think things might be a little more complicated.  You're on your own if that's all you've got.  I don't know if some adaptation of my setup can be devised that would work.  That's all up to you to figure out.

Something I have found with screen recording is that things can get tricky for recording sound.  It's pretty easy to get sound from 2 sources in your recording, which is pretty useless.  You'll get the sound track of what you want plus you'll get the sound from anything else you might be watching on your other monitor while you're recording.  In addition, doing things like Alt+Tab to switch tasks causes momentary interruption of the video track of your recording.  So for a perfect recording, you're probably going to have to pretty much not use your system while the recording is in progress.  It is always preferable to do downloads.  But when downloads aren't possible, like when the content is protected by DRM, you're stuck with what I've described here.

**Reply by sortedcord** [reply]
Its true that my system is unusable if I decide to record lectures.. I might just have to use keydive or something to obtain the keys and then make my way through that. Although the setup which you have described is quite interesting. And yeah my system is relatively old which might explain why im able to record with obs (6th gen intel)

**Reply by GrampaWildWilly** [reply]
I don't have my old W7 box any more.  But it seems to me Intel wasn't even using the "generation" terminology for their CPUs when I got it.  My new (now 1 year old) W11 box has a 13th generation Intel CPU, so I've advanced to the point of not being able to do what used to work . . . except now I've added these 2 bits of hardware that get around the problem.  But even with this, recording something means basically don't use the computer while that's going on.  I would typically record something while I'm asleep or not home.  If I decide to record something while I want to use the computer, I do the Alt+Tab thing as quickly as I can in order to minimize the dropouts of the video in the recording.  If I'm recording something like hockey or golf, a momentary dropout is not a big deal.  But if I were recording a lecture, like you are, or a movie or something I bought, I would definitely do it overnight or while I'm not even home.  Plus I would not be trying to watch or listen to anything else while the recording was in progress.  I think there's a way in OBS to isolate what audio it records so you don't record 2 audio tracks into the same video.  But I believe there are still cases in which that isolation is not possible.  So I just don't try to overdo the multitasking while I'm doing a recording.

I'd never heard of Keydive before you mentioned it.  So I looked it up.  It says it's for use on Android devices.  How does that help you on a Windows PC?  On further inspection, I see that it requires you to root your phone.  I don't know where you live but in this country (USA), the phone service providers & phone manufacturers have conspired to make it impossible to root phones any more.  I had a Galaxy S6 & old as that was, the version of Android on that (I think it might have been 8 but don't quote me on that) did not have the setting for enabling replacement of the bootloader.  I know what the setting is because I saw it in YouTube videos I watched a couple of years ago.  My phone simply didn't have that setting.  It is my understanding that in other countries, you can still root your phone.  But I'm not sure what releases of Android still allow this.  I know my Galaxy S24FE Android 14 phone doesn't have the switch for enabling replacement of the bootloader.  Our world of freedom seems to be a world in which corporations are free to prevent users from doing what they'd like to do.

But even if I could root my phone, I think it's much easier to just get the 2 bits of hardware I got.  To record DRM content now is pretty seamless.  It's as if the capability is built into Windows, even though it isn't.

**Reply by devinschumacher** [reply]
> Supposedly OBS could record DRM video but then they changed it so that even recording doesn't work. Maybe you have an older version of the software. The idea of DRM is to allow the video to play online even in PiP mode but prevent anyone from keeping a copy of it locally on your computer. The video player decrypts the stream so that the viewer can obviously watch it. But if you attempt to download it, it's still encrypted therefore you get a result like this. Totally unusable file.
> 
> Recording works because it's like using a cassette tape to record a song directly from the radio station. It's captures the video as an unencrypted stream like on the webpage. I've read somewhere online that hardware capture cards can also record video.
> 
> ![drm video](https://private-user-images.githubusercontent.com/160297904/393469683-14544b3a-90ab-4ebe-adc8-620fdab32f61.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM5NzMxOTQsIm5iZiI6MTczMzk3Mjg5NCwicGF0aCI6Ii8xNjAyOTc5MDQvMzkzNDY5NjgzLTE0NTQ0YjNhLTkwYWItNGViZS1hZGM4LTYyMGZkYWIzMmY2MS5QTkc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxMlQwMzA4MTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02M2NkYjVmNDBjODRkN2ZhMjM1NWQ0OTlkYjA2YWUxMmJlYmFmMjVhMzMyYjUyYWRlNDI2MTc0MzRiYjM3MDFlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.mvjZD59ExNYvNu5L3pkhOzvJhwpuzh8sUrzr6-kILDQ)



try this method --- disable acceleration first.

walkthrough: https://www.youtube.com/watch?v=dowl67QPLlg

**Reply by sortedcord** [reply]
> I don't have my old W7 box any more. But it seems to me Intel wasn't even using the "generation" terminology for their CPUs when I got it. My new (now 1 year old) W11 box has a 13th generation Intel CPU, so I've advanced to the point of not being able to do what used to work . . . 

I'm using swaywm on arch linux. Ended up creating a dummy output screen and a dummy output device. Moved Firefox to the dummy screen and redirected its audio to the dummy output device and recorded that with obs. Can now record the lectures while using my system no problem. Yes, its whack but I can at least use those videos for offline viewing since I use an old ipad which is still a very capable device, just that udemy is no longer supported on it...

**Reply by GrampaWildWilly** [reply]
Interesting.  So it appears you have tools on Linux for emulating in software exactly what I've done in hardware.  Maybe there's similar software available in Windows.  That's a moot point for me since I already have the 2 bits of hardware.  But somebody else might be motivated to investigate this & report back here.
